sklearn/linear_model/logistic.py

```diff
@@ -1203,7 +1203,13 @@ def fit(self, X, y, sample_weight=None):
             raise ValueError("Tolerance for stopping criteria must be "
                              "positive; got (tol=%r)" % self.tol)
 
-        X, y = check_X_y(X, y, accept_sparse='csr', dtype=np.float64,
+        # setup the default dtype
+        _dtype = np.float64
+        if self.solver in ['newton-cg'] \
+                and isinstance(X, np.ndarray) and X.dtype in [np.float32]:
+            _dtype = np.float32
```

Comment:
check_X_y can take a list of acceptable dtypes as a dtype argument. I think that using this feature would be a better way of writing this code. The code would be something like 

<pre>
if self.solver in ['newtown-cg']:
   _dtype = [np.float64, np.float32]
</pre>